From e165c4981b5f8c9a48e9f8bf267a001308f29d4b Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Wed, 28 Nov 2018 12:49:40 -0800
Subject: [PATCH] Various fixes related to Clear Linux (#431)

* Also include base libraries for AVX2 and AVX512 builds

glibc supports searching for libraries in subdirs according to the CPU
features. On x86-64, those are "haswell" and "avx512_1", as in:

$ ldd /usr/bin/icecc | grep haswell
        libstdc++.so.6 => /usr/lib64/haswell/avx512_1/libstdc++.so.6
        libc.so.6 => /usr/lib64/haswell/libc.so.6
        libm.so.6 => /usr/lib64/haswell/avx512_1/libm.so.6

With this change, we'll include the base library or libraries in the
environment. In the system above, we'll include /usr/lib64/libm.so.6,
/usr/lib64/haswell/libm.so.6 in addition to the libm listed in ldd's
output.

* Fix finding ld.so.conf on Clear Linux

Clear Linux has a "stateless" /etc, meaning that all files there are
supposed to belong to the user. The ld.so.cache file is instead in
/var/cache/ldconfig.

* Fix warning printed when /lib or /lib64 are symlinks

icecc-create-env was printing:
[...]
adding file /usr/bin/objcopy
adding file /etc/ld.so.conf=/tmp/icecc_ld_so_confTsSNqs
cp: -r not specified; omitting directory '/lib'
cp: -r not specified; omitting directory '/lib64'
creating d7801039695fbb794e026bdaf287276c.tar.gz

The -p option means to preserve ownership, timestamps, etc., but doesn't
tell cp to operate on the symlink itself. The -P option does.
---
 client/icecc-create-env.in | 26 ++++++++++++++++++++++++--
 1 file changed, 24 insertions(+), 2 deletions(-)

diff --git a/client/icecc-create-env.in b/client/icecc-create-env.in
index ce9c6fe..a0227cc 100755
--- a/client/icecc-create-env.in
+++ b/client/icecc-create-env.in
@@ -109,6 +109,7 @@ add_file ()
 		#         libc.so.6 => /lib/tls/libc.so.6 (0xb7e81000)
 		#         /lib/ld-linux.so.2 (0xb7fe8000)
 		# covering both situations ( with => and without )
+           local lib
            for lib in $(ldd "$path" | sed -n 's,^[^/]*\(/[^ ]*\).*,\1,p'); do
 	      test -f "$lib" || continue
 	      # Check wether the same library also exists in the parent directory,
@@ -116,6 +117,23 @@ add_file ()
 	      local baselib=$(echo "$lib" | sed 's,\(/[^/]*\)/.*\(/[^/]*\)$,\1\2,')
 	      test -f "$baselib" && lib=$baselib
               add_file "$lib"
+
+              # Add the non-haswell and non-avx512_1 libraries too
+              case "$lib" in
+                 */haswell/*|*/avx512_1/*)
+                    ;;
+                 *)
+                    continue
+                    ;;
+              esac
+              local lib_non_avx512=$(echo "$lib" | sed s,/avx512_1/,/,)
+              local lib_non_hsw=$(echo "$lib_non_avx512" | sed s,/haswell/,/,)
+              if [ "$lib" != "$lib_non_avx512" ] && [ -f "$lib_non_avx512" ]; then
+                add_file "$lib_non_avx512"
+              fi
+              if [ "$lib" != "$lib_non_hsw" ] && [ -f "$lib_non_hsw" ]; then
+                add_file "$lib_non_hsw"
+              fi
            done
         fi
     elif test "$is_darwin" = 1; then
@@ -434,7 +452,7 @@ fi
 
 # special case for weird multilib setups
 for dir in /lib /lib64 /usr/lib /usr/lib64; do
-    test -L $dir && cp -p $dir $tempdir$dir
+    test -L $dir && cp -Pp $dir $tempdir$dir
 done
 
 new_target_files=
@@ -458,7 +476,11 @@ done
 if test -x /sbin/ldconfig -a "$is_linux" = 1; then
    mkdir -p $tempdir/var/cache/ldconfig
    /sbin/ldconfig -r $tempdir
-   new_target_files="$new_target_files etc/ld.so.cache"
+   for candidate in etc var/cache/ldconfig; do
+        test -e $tempdir/$candidate/ld.so.cache || continue;
+        new_target_files="$new_target_files $candidate/ld.so.cache"
+        break
+   done
 fi
 
 md5sum=NONE
-- 
2.19.2

